Speed_Controller/src/main.c

```diff
@@ -0,0 +1,167 @@
+	CAN_drive.data[2] = v.chars[2];
+	CAN_drive.data[3] = v.chars[3];
+
+	//set velocity to 0 for regen messages
+	CAN_regen.data[0] = 0;
+	CAN_regen.data[1] = 0;
+	CAN_regen.data[2] = 0;
+	CAN_regen.data[3] = 0;
+	
+	//set initial 
+	CAN_drive.data[4] = 0;
+	CAN_drive.data[5] = 0;
+	CAN_drive.data[6] = 0;
+	CAN_drive.data[7] = 0;
+	
+	CAN_regen.data[4] = 0;
+	CAN_regen.data[5] = 0;
+	CAN_regen.data[6] = 0;
+	CAN_regen.data[7] = 0;
+		
+	RCC->APB2ENR |= 0x1 << 4;  
+	GPIOC->CRH |= 0x4;
+	GPIOC->CRL |= 0x4 << 24;
+	
+	volatile uint16_t encoder_reading;
+	volatile uint16_t ADC_reading;
+	uint16_t old_ADC_reading = 0x000;
+	uint16_t old_encoder_reading = 0x000;
+	uint8_t regen_enabled;
+	uint8_t current_direction;
+	uint8_t previous_direction = 0;
+	
+	while(1) 
+	{
+		
+		encoder_reading = EncoderRead();
+		ADC_reading = ReadADC();
+		
+		regen_enabled = (GPIOC->IDR >> 8) & 0x1UL;
+		current_direction = (GPIOC->IDR >> 6) & 0x1UL;
+		
+		if (current_direction != previous_direction)
+		{
+			v.float_var = v.float_var * -1;
+			
+			CAN_drive.data[0] = v.chars[0];
+			CAN_drive.data[1] = v.chars[1];
+			CAN_drive.data[2] = v.chars[2];
+			CAN_drive.data[3] = v.chars[3];			
+			
+			previous_direction = current_direction;
+		}
+		
+		//If regen is enabled AND ADC count changed, send new regen CAN message and restart timer
+		if( regen_enabled && (old_ADC_reading != ADC_reading) )
+		{
+			
+			StopTimer();
+			
+			u.float_var = (float) ADC_reading / ADC_ZERO_THRESHOLD;
+			
+			CAN_regen.data[4] = u.chars[0];
+			CAN_regen.data[5] = u.chars[1];
+			CAN_regen.data[6] = u.chars[2];
+			CAN_regen.data[7] = u.chars[3];
+			
+			CANSend(&CAN_regen);
+			
+			old_ADC_reading = ADC_reading;
+			old_encoder_reading = encoder_reading;
+			
+			RestartTimer();
+		}		
+		//if encoder count changed, send new drive CAN message and restart timer
+		else if(old_encoder_reading != encoder_reading && (ADC_reading == 0 || !regen_enabled))
+		{
+			
+			u.float_var = (float)( (float) encoder_reading/PEDAL_MAX);
+			
+			//Any overflow should be treated as zero current to the motor
+			if (u.float_var > 1.0 || u.float_var < 0.0)
+			{
+				u.float_var = 0.0;
+			}
+			else
+			{
+				old_ADC_reading = ADC_reading;
+				old_encoder_reading = encoder_reading;				
+			}
+			
+			CAN_drive.data[4] = u.chars[0];
+			CAN_drive.data[5] = u.chars[1];
+			CAN_drive.data[6] = u.chars[2];
+			CAN_drive.data[7] = u.chars[3];
+			
+			CANSend(&CAN_drive);
+		
+			RestartTimer();
+		
+		}		
+		//If a timeout occured, send the previously sent CAN drive message
+		else if(timeoutFlag == TRUE)
+		{	
+			
+			//If the driver is holding regen at a positive value, send the previous regen message
+			if(ADC_reading > 0 && regen_enabled)
+			{	
+				CANSend(&CAN_regen);
+			}
+			//If the driver is not braking, send the previous CAN drive message.
+			else
+			{	
+				CANSend(&CAN_drive);
+			}
+			
+			timeoutFlag = FALSE;
+			
+			old_ADC_reading = ADC_reading;
```

Comment:
why is this in each branch? why can't they just be outside the cases and at the end of the while loop?